Dana/hf-alb-docs (#2276)
huggingface/blog: doc_aug_hf_alb.md

---
title: "Introducing TextImage Augmentation for Document Images" 
thumbnail: "/blog/assets/185_albumentations/thumbnail.png"
authors:
- user: danaaubakirova
- user: Ternaus
guest: True
---
# Introducing Multimodal TextImage Augmentation for Document Images

In this blog post, we provide a tutorial on how to use a new data augmentation technique for document images, developed in collaboration with Albumentations AI.

## Motivation

> The need for effective data augmentation techniques for document images became evident when addressing challenges in fine-tuning models with limited datasets. A common concern is that typical image transformations, such as resizing, blurring, or changing background colors, can negatively impact text extraction accuracy.

Reviewer: I'd try to recontextualize a bit the posting wrt to the wave of VLMs:
```suggestion
Visual Language Models (VLMs) have an immense range of application, but they often need to be fine-tuned to a range of specific use-cases, in particular for datasets containing document images, i.e., images with high textual content. And we want a model to learn to read properly, which is made difficult in the most frequent of cases: where data is missing.

Hence, the need for **effective data augmentation** techniques for document images became evident when addressing challenges in fine-tuning models with limited datasets. A common concern is that typical image transformations, such as resizing, blurring, or changing background colors, can negatively impact text extraction accuracy.
```

Author: Yes, I think this is a good point. It is better when the text-richness of documents and their tole in fine-tuning and pertaining VLMs is clearly articulated.

![png](assets/185_albumentations/quote2.png)

> We recognized the need for data augmentation techniques that preserve the integrity of the text while augmenting the dataset. Such data augmentation can facilitate the easy generation of new documents or the modification of existing ones, while preserving the quality of the text.

Reviewer: ```suggestion
We recognized the need for data augmentation techniques that preserve the integrity of the text while augmenting the dataset. Such data augmentation can facilitate generation of new documents or modification of existing ones, while preserving their text quality.
```

## Introduction

> To address the need, we are introducing a new data augmentation pipeline developed in collaboration with Albumentations AI. This pipeline is designed to handle both images and text within them, providing a comprehensive solution for document images. We refer to this type of data augmentation as *multimodal*, as it modifies both the image content and the text annotations.

Reviewer: Also maybe insert link to albumentations here? 
```suggestion
To address this need, we introduce a **new data augmentation pipeline** developed in collaboration with Albumentations AI. This pipeline handles both images and text within them, providing a comprehensive solution for document images. This class of data augmentation is *multimodal* as it modifies both the image content and the text annotations simultaneously.
```

Author: added the link

> As discussed in a previous [blog post](https://huggingface.co/blog/danaaubakirova/doc-augmentation), our goal is to test the hypothesis that integrating augmentations on both text and images during pretraining is effective. Detailed parameters and use case illustrations can be found on the  [Albumentations AI Documentation](https://albumentations.ai/docs/examples/example_textimage/?h=textimage). Albumentations AI enables the dynamic design of these augmentations and their integration with other types of augmentations.

Reviewer: ```suggestion
As discussed in a previous [blog post](https://huggingface.co/blog/danaaubakirova/doc-augmentation), our goal is to test the hypothesis that integrating augmentations on both text and images during pretraining of VLMs is effective. Detailed parameters and use case illustrations can be found on the  [Albumentations AI Documentation](https://albumentations.ai/docs/examples/example_textimage/?h=textimage). Albumentations AI enables the dynamic design of these augmentations and their integration with other types of augmentations.
```

# Method 

> To augment document images, we begin by randomly selecting lines within the document. A hyperparameter `fraction_range` controls the fraction of bounding box to be modified.

Reviewer: ```suggestion
To augment document images, we begin by randomly selecting lines within the document. A hyperparameter `fraction_range` controls the bounding box fraction to be modified.
```

```python
transform = A.Compose([A.TextImage(font_path=font_path, p=1, augmentations=["deletion"], clear_bg=True, font_color = 'red', fraction_range = (0.5,0.8), font_size_fraction_range=(0.8, 0.9))])
```

```python
transformed = transform(image=image, textimage_metadata=metadata)
```

```python
visualize(transformed['image'])
```

    
![png](assets/185_albumentations/output_25_0.png)
    

## Random Insertion

```python
!pip install nltk
```

    Requirement already satisfied: nltk in /fsx/dana_aubakirova/miniconda/envs/myenv/lib/python3.11/site-packages (3.8.1)
    Requirement already satisfied: click in /fsx/dana_aubakirova/miniconda/envs/myenv/lib/python3.11/site-packages (from nltk) (8.1.7)
    Requirement already satisfied: joblib in /fsx/dana_aubakirova/miniconda/envs/myenv/lib/python3.11/site-packages (from nltk) (1.3.2)
    Requirement already satisfied: regex>=2021.8.3 in /fsx/dana_aubakirova/miniconda/envs/myenv/lib/python3.11/site-packages (from nltk) (2023.12.25)
    Requirement already satisfied: tqdm in /fsx/dana_aubakirova/miniconda/envs/myenv/lib/python3.11/site-packages (from nltk) (4.65.0)

```python
import nltk

nltk.download('stopwords')

> from nltk.corpus import stopwords

Reviewer 2: More preparation which I would handle at the start

```

    [nltk_data] Downloading package stopwords to
    [nltk_data]     /admin/home/dana_aubakirova/nltk_data...
    [nltk_data]   Package stopwords is already up-to-date!

```python
stops = stopwords.words('english')
```

```python
transform = A.Compose([A.TextImage(font_path=font_path, p=1, augmentations=["insertion"], stopwords = stops, clear_bg=True, font_color = 'red', fraction_range = (0.5,0.8), font_size_fraction_range=(0.8, 0.9))])
```

```python
transformed = transform(image=image, textimage_metadata=metadata)
```

```python
visualize(transformed['image'])
```

> 

Reviewer 2: all of this seems like something to explain. I know nltk but for someone who doesn't this code is hard to follow.

Author: Good point. I tried to explain what are stop words.

    
![png](assets/185_albumentations/output_32_0.png)
    

## Can we combine with other transformations?

```python
transform_complex = A.Compose([A.TextImage(font_path=font_path, p=1, augmentations=["insertion"], stopwords = stops, clear_bg=True, font_color = 'red', fraction_range = (0.5,0.8), font_size_fraction_range=(0.8, 0.9)),
                               A.PlanckianJitter(p=1),
                               A.Affine(p=1)
                              ])
```

```python
transformed = transform_complex(image=image, textimage_metadata=metadata)
```

```python
visualize(transformed["image"])

> ```

Reviewer 2: more things to explain, and code to put together. Just having a title is not enough 😅

Author: Added explanation.

      'text': "to In its current form, the guidance does not they hers this adequately address what their is needed to y didn't",
      'original_text': 'In its current form, the guidance does not adequately address what is needed to',
      'bbox_index': 29,
      'font_color': 'red'},
     {'bbox_coords': (375, 1741, 2171, 1789),
      'text': 'Corporation delivered a presentation entitled "Brand these Industry Perspective nor then on the if d',
      'original_text': 'Corporation delivered a presentation entitled "Brand Industry Perspective on the',
      'bbox_index': 20,
      'font_color': 'red'},
     {'bbox_coords': (562, 583, 1987, 631),
      'text': 'PUBLIC MEETING ON PRE-MARKET is hasn EVALUATION OF than',
      'original_text': 'PUBLIC MEETING ON PRE-MARKET EVALUATION OF',
      'bbox_index': 7,
      'font_color': 'red'},
     {'bbox_coords': (525, 2375, 2173, 2421),
      'text': "Emphasized that it couldn don't won't is critical that the she generic shouldn product is weren no less abuse-deterrent",
      'original_text': 'Emphasized that it is critical that the generic product is no less abuse-deterrent',
      'bbox_index': 26,

>       'font_color': 'red'}]

Reviewer 2: This is just printed as text. Are you sure that you want to do that? How will it look like in the blog?

Reviewer: Maybe just one element of this dictionary is enough - no need for it to be copy-pastable either, so an image containing that would more readable/enticing to read, for instance on the left the modified image cropped to the line of interest, and on the right an image of the dictionary entry of interest?

## Synthetic Data Generation

This augmentation method can be extended to the generation of synthetic data, as it enables the rendering of text on any background or template.

```python
template = cv2.imread('template.png')
image_template = cv2.cvtColor(template, cv2.COLOR_BGR2RGB)
```

```python
transform = A.Compose([A.TextImage(font_path=font_path, p=1, clear_bg=True, font_color = 'red', font_size_fraction_range=(0.5, 0.7))])
```

```python
metadata = [{
    "bbox": [0.1, 0.4, 0.5, 0.48],
    "text": "Some smart text goes here.",
}, {
    "bbox": [0.1, 0.5, 0.5, 0.58],
    "text": "Hope you find it helpful.",
}]
```

```python
transformed = transform(image=image_template, textimage_metadata=metadata)
```

```python
visualize(transformed['image'])

> ```

Reviewer 2: all together

    
![png](assets/185_albumentations/output_46_0.png)
    

```python

> ```

Reviewer 2: empty python blocK?

Author: removed
